File: ref/bfs.cpp
#include <queue>
#include <vector>
#include <iostream>

using namespace std;

int number;
int c[7];
vector<int> a[8];

void bfs(int start) {
    queue <int> q;
    q.push(start);
    c[start] = true;
    while (!q.empty()) {
        int x = q.front();
        q.pop();
        // cout << x << ' ';
        for (int i = 0; i < a[x].size(); i++) {
            int y = a[x][i];
            cout << "백터 " << x << " 와 연결된 노드 " << y << " - " ;
            if (!c[y]) {
                cout << "방문하지 않은 노드 " << endl;
                q.push(y);
                c[y] = true;
            }
            else 
                cout << "방문한 노드 "<< endl;
        }
    }
}

int main(void)
{
    a[1].push_back(2);
    a[2].push_back(1);

    a[1].push_back(3);
    a[3].push_back(1);

    a[2].push_back(4);
    a[4].push_back(2);

    a[2].push_back(5);
    a[5].push_back(2);

    a[3].push_back(6);
    a[6].push_back(3);

    a[3].push_back(7);
    a[7].push_back(3);

    bfs(1);

}

File: ref/bfs.js
function bfs(graph, start, visited) {
  const queue = [start];
  visited.add(start);

  while (queue.length > 0) {
    const student = queue.shift(); // 큐에서 학생을 꺼냄
    for (const friend of graph[student]) {
      if (!visited.has(friend)) {
        visited.add(friend);
        queue.push(friend); // 친구를 큐에 추가
      }
    }
  }
}

function minFriendships(n, friendships) {
  const graph = new Map();

  // 그래프 생성
  for (const [a, b] of friendships) {
    if (!graph.has(a)) graph.set(a, []);
    if (!graph.has(b)) graph.set(b, []);
    graph.get(a).push(b);
    graph.get(b).push(a);
  }

  const visited = new Set();
  let components = 0;

  // 각 학생에 대해 BFS 수행
  for (let student = 1; student <= n; student++) {
    if (!visited.has(student)) {
      bfs(graph, student, visited);
      components++; // 새로운 연결 요소 발견
    }
  }

  return components;
}

// 예시 사용
const n = 5;
const friendships = [
  [1, 2],
  [2, 3],
  [4, 5],
];
console.log(minFriendships(n, friendships)); // 결과 출력
